dbms/src/Flash/Mpp/MPPTask.cpp

```diff
@@ -258,35 +258,46 @@ void MPPTask::registerTunnels(const mpp::DispatchTaskRequest & task_request)
-            auto exchange_receiver = std::make_shared<ExchangeReceiver>(
-                std::make_shared<GRPCReceiverContext>(
-                    executor.exchange_receiver(),
-                    dag_context->getMPPTaskMeta(),
-                    context->getTMTContext().getKVCluster(),
-                    context->getTMTContext().getMPPTaskManager(),
-                    context->getSettingsRef().enable_local_tunnel,
-                    context->getSettingsRef().enable_async_grpc_client),
-                executor.exchange_receiver().encoded_task_meta_size(),
-                context->getMaxStreams(),
-                log->identifier(),
-                executor_id,
-                executor.fine_grained_shuffle_stream_count(),
-                context->getSettingsRef());
-
-            if (status != RUNNING)
-                throw Exception(
-                    "exchange receiver map can not be initialized, because the task is not in running state");
-
-            receiver_set_local->addExchangeReceiver(executor_id, exchange_receiver);
-        }
-        return true;
-    });
+    try
+    {
+        dag_context->dag_request.traverse([&](const tipb::Executor & executor) {
+            if (executor.tp() == tipb::ExecType::TypeExchangeReceiver)
+            {
+                assert(executor.has_executor_id());
+                const auto & executor_id = executor.executor_id();
+                // In order to distinguish different exchange receivers.
+                auto exchange_receiver = std::make_shared<ExchangeReceiver>(
+                    std::make_shared<GRPCReceiverContext>(
+                        executor.exchange_receiver(),
+                        dag_context->getMPPTaskMeta(),
+                        context->getTMTContext().getKVCluster(),
+                        context->getTMTContext().getMPPTaskManager(),
+                        context->getSettingsRef().enable_local_tunnel,
+                        context->getSettingsRef().enable_async_grpc_client),
+                    executor.exchange_receiver().encoded_task_meta_size(),
+                    context->getMaxStreams(),
+                    log->identifier(),
+                    executor_id,
+                    executor.fine_grained_shuffle_stream_count(),
+                    context->getSettingsRef());
+
+                receiver_set_local->addExchangeReceiver(executor_id, exchange_receiver);
+
+                if (status != RUNNING)
+                    throw Exception(
+                        "exchange receiver map can not be initialized, because the task is not in running state");
+            }
+            return true;
+        });
+    }
+    catch (...)
+    {
+        std::unique_lock lock(mtx);
+        if (status != RUNNING)
+            throw Exception("exchange receiver map can not be initialized, because the task is not in running state");
+        receiver_set = std::move(receiver_set_local);
```

Comment:
Why not just remove `receiver_set_local` and use `receiver_set` directly

Reply:
make `MPPReceiveSet` thread safe

Reply:
keep MPPReceiverSet atomic

Reply:
But it doesn't matter if it's not atomic.
- The mpp task will not execute until initExchangeReceiver is completed
- When canceling, MPPReceiverSet can be incomplete

Reply:
IMO, when canceling, MPPTaskManager is responsible for release MPPReceiverSet resources. If new exchange receiver added after MPPTaskManager released work done, it will be problematic.

Reply:
But MPPReceveSet will be destructed and then eventually be closed

Reply:
I think here both ways can work "correctly" if details are handled correctly. And we can distiguish "release resource" and destruct, mpp manager is responsible for release resource when canceling.

Reply:
ok